File: assets/js/preferences.js
// Persisted presentation preferences. The server renders the cookie-backed
// state on first byte; this enhancement changes the root attributes, keeps the
// controls' state current, and re-syncs both after a back/forward-cache
// restore revives HTML older than the cookies.
//
// Two places set these choices — the controls in the header, and the page that
// names each one and says what it does — and both leave through the same
// doors below, so there is one way a choice reaches this browser rather than
// one per control. A write is read back before it is believed: storage this
// address may not use is refused silently, and a choice nothing kept must not
// be left looking chosen.
//
// Language is not handled here: its words are the server's, so its control is
// a plain form the server answers with a redirect. This module only checks,
// after a cache restore, that the revived document still speaks the language
// the cookie names — and asks for a fresh page when it does not, since no
// script can retranslate a rendered document.
export function initPreferences() {
  const root = document.documentElement;

  function readCookie(name) {
    for (const part of document.cookie.split(';')) {
      const eq = part.indexOf('=');
      if (eq > -1 && part.slice(0, eq).trim() === name) {
        return part.slice(eq + 1).trim();
      }
    }
    return null;
  }

  function setPreference(name, value) {
    root.dataset[name] = value;
    document.cookie = `yomihon_${name}=${value};path=/;max-age=31536000;samesite=lax`;
  }

  function setSingleKeyShortcuts(value) {
    root.dataset.singleKeyShortcuts = value;
    document.cookie = `yomihon_shortcuts=${value};path=/;max-age=31536000;samesite=lax`;
  }

  const themeToggle = document.querySelector('[data-theme-toggle]');
  const textsizeToggle = document.querySelector('[data-textsize-toggle]');
  const rubyToggle = document.querySelector('[data-ruby-toggle]');
  const shortcutsToggle = document.querySelector('[data-single-key-shortcuts-toggle]');

  // The size names come from the control itself, where the server wrote them
  // in the page's language; a copy here would be a second dictionary.
  function textsizeLabel(size) {
    return textsizeToggle?.dataset[{ m: 'labelM', l: 'labelL', xl: 'labelXl' }[size]];
  }

  // What the reader currently sees: their stored choice when one is on the
  // root, otherwise whichever theme the system preference painted.
  function effectiveTheme() {
    return (
      root.dataset.theme ||
      (matchMedia('(prefers-color-scheme: dark)').matches ? 'dark' : 'light')
    );
  }

  // Anything drawn in the theme's own colours, rather than styled by the
  // stylesheet, has to be told when the theme moves. One listener, because one
  // thing is drawn that way.
  let themeChanged = () => {};

  // The server stamps this control's pressed state from the stored choice,
  // which is all it can see: prefers-color-scheme never reaches it. With no
  // choice stored and a dark system preference the page paints dark and the
  // attribute says otherwise. So the control is told from here, from what the
  // reader is actually looking at — on arrival, after a write, and when the
  // system moves underneath a reader who stored nothing.
  function markThemePressed() {
    themeToggle?.setAttribute('aria-pressed', String(effectiveTheme() === 'dark'));
  }

  // The only place the reader's theme reaches the root element. A choice they
  // made is stored as well; the rewrite from the cookie after a cache restore
  // is not, and may have no value to write at all. Both leave through the same
  // announcement, so a write site added later cannot leave a drawing a frame
  // behind the page by not thinking to mention itself.
  function writeTheme(choice, stored) {
    const before = effectiveTheme();
    if (stored) {
      setPreference('theme', choice);
    } else if (choice) {
      root.dataset.theme = choice;
    } else {
      delete root.dataset.theme;
    }
    const after = effectiveTheme();
    markThemePressed();
    // A write that left the theme where it was is not news. The cookie read
    // after a cache restore usually names the theme the page is already
    // showing, and announcing that would redraw every diagram on the page each
    // time the reader stepped back to it.
    if (after !== before) themeChanged();
  }

  // Nothing has moved yet: this is the server's own stamp going through the
  // same door, so the control agrees with what the reader is looking at from
  // the first paint rather than only after they touch it.
  writeTheme(root.dataset.theme || null, false);

  // A reader who stored no choice is following the system, and the system can
  // change while they are reading: the stylesheet repaints on its own, but the
  // control that reports the theme and anything drawn in the theme's colours
  // are told here. A stored choice fixes the theme, so the same change moves
  // nothing and says nothing.
  matchMedia('(prefers-color-scheme: dark)').addEventListener('change', () => {
    if (root.dataset.theme) return;
    markThemePressed();
    themeChanged();
  });

  // The size reaches the root, the cookie, and the control that reports it
  // through one door. The control cycles three sizes, so its state lives in its
  // accessible name rather than in aria-pressed, and a write that moved the
  // size without renaming it would leave a reader who cannot see the type being
  // told the wrong thing by the very control that names it.
  function writeTextSize(size) {
    setPreference('textsize', size);
    textsizeToggle?.setAttribute('aria-label', textsizeLabel(size));
  }

  // Furigana and the single-key keys are switches, and their controls report
  // which state they are in, so each value and the control saying it move
  // together for the same reason the size does.
  function writeRuby(value) {
    setPreference('ruby', value);
    rubyToggle?.setAttribute('aria-pressed', String(value === 'on'));
  }

  function writeShortcuts(value) {
    setSingleKeyShortcuts(value);
    if (shortcutsToggle) shortcutsToggle.checked = value === 'on';
  }

  textsizeToggle?.addEventListener('click', () => {
    writeTextSize({ m: 'l', l: 'xl', xl: 'm' }[root.dataset.textsize] || 'l');
  });
  themeToggle?.addEventListener('click', () => {
    // The flip starts from what the reader sees: with no stored choice the
    // page may already be dark from the system, and the first press must then
    // choose light rather than restate dark.
    writeTheme(effectiveTheme() === 'dark' ? 'light' : 'dark', true);
  });
  rubyToggle?.addEventListener('click', () => {
    writeRuby(root.dataset.ruby === 'off' ? 'on' : 'off');
  });
  shortcutsToggle?.addEventListener('change', (event) => {
    const value = event.currentTarget.checked ? 'on' : 'off';
    writeShortcuts(value);
  });

  // The page a reader sets the reading on. Its choices are the same six, so
  // they leave through the doors above rather than through a second way of
  // storing the same thing. The form those radios sit in stays underneath,
  // unchanged and unused here: it is what applies a choice for a browser
  // running none of this, and it is still how the interface language is
  // applied for every browser, because the words on a rendered page are
  // written by the server and no script can rewrite them.
  const settingsForm = document.querySelector('[data-prefs]');

  // What each choice does to this browser when it is picked. A choice missing
  // from here is one nothing on this side can carry out, and is asked of the
  // server instead.
  const applyChoice = {
    theme(value, stored) {
      if (stored) {
        writeTheme(value, true);
        return;
      }
      // Following the system is held by the absence of a stored value, so
      // picking it removes one rather than storing a word meaning "none" —
      // the same deletion the server writes for this option.
      document.cookie = 'yomihon_theme=;path=/;max-age=0;samesite=lax';
      writeTheme(null, false);
    },
    textsize: writeTextSize,
    font(value) {
      setPreference('font', value);
    },
    ruby: writeRuby,
    shortcuts: writeShortcuts,
  };

  // Where each choice currently stands, so one this browser refuses can be put
  // back where it was instead of left sitting on a value nothing stored.
  const standing = new Map();

  function settingsFields() {
    return settingsForm ? settingsForm.querySelectorAll('[data-pref-field]') : [];
  }

  function optionsOf(fieldset) {
    return [...fieldset.querySelectorAll('input[type="radio"]')];
  }

  // Whether picking this option stores a value. The page marks the options the
  // server would accept into the cookie; the one it does not mark is held by
  // that cookie's absence.
  function storesAValue(option) {
    return option.dataset.prefStores !== undefined;
  }

  // Each choice lands in the cookie its own name spells, which is the shape
  // the server's table is held to.
  function storedValue(name) {
    return readCookie(`yomihon_${name}`);
  }

  function applyOption(name, option) {
    applyChoice[name](option.value, storesAValue(option));
  }

  for (const fieldset of settingsFields()) {
    const current = fieldset.querySelector('input[type="radio"]:checked');
    if (current) standing.set(fieldset.dataset.prefField, current);
  }

  settingsForm?.addEventListener('change', (event) => {
    const option = event.target;
    if (!(option instanceof HTMLInputElement) || option.type !== 'radio') return;
    const fieldset = option.closest('[data-pref-field]');
    if (!fieldset) return;
    const name = fieldset.dataset.prefField;
    const status = fieldset.querySelector('[data-pref-status]');
    if (!applyChoice[name]) {
      // Asking the server, because only the server can rewrite the words it
      // wrote — and asking in a way that comes back to this page, since a
      // reader who is still choosing has more to set. Nothing they have picked
      // so far is waiting on this, so nothing is lost by leaving.
      const next = settingsForm.querySelector('input[name="next"]');
      if (next) next.value = settingsForm.dataset.prefsSettings;
      settingsForm.requestSubmit();
      return;
    }
    applyOption(name, option);
    // A browser that refuses this address its storage takes the write in
    // silence, leaving the page showing a choice nothing kept. So the write is
    // read back before it is believed, and a choice that did not survive is
    // put back and said out loud rather than left looking chosen.
    const kept = storedValue(name);
    if (storesAValue(option) ? kept !== option.value : kept !== null) {
      const previous = standing.get(name);
      if (previous) {
        applyOption(name, previous);
        previous.checked = true;
      }
      if (status) status.textContent = status.dataset.prefRefused;
      return;
    }
    standing.set(name, option);
    // A sentence about a choice that could not be kept is answered by the same
    // choice being kept, and has to go when it is.
    if (status) status.textContent = '';
  });

  // A revived page carries the radios as the reader left them, while the
  // cookies may have moved on — a size chosen from the control above on a later
  // page arrives back here with the old one still marked. The options
  // themselves carry which values may be stored and which of them answers for
  // storing none, so the cookie is read through the list the server rendered
  // rather than through a second copy of its rules.
  function syncSettingsChoices() {
    for (const fieldset of settingsFields()) {
      const name = fieldset.dataset.prefField;
      if (!applyChoice[name]) continue;
      const kept = storedValue(name);
      const options = optionsOf(fieldset);
      const match =
        options.find((option) => storesAValue(option) && option.value === kept) ||
        options.find((option) => option.dataset.prefUnset !== undefined);
      if (!match) continue;
      match.checked = true;
      standing.set(name, match);
    }
  }

  // A back/forward-cache restore revives the document exactly as it left,
  // while the cookies may have moved on — a theme chosen on the next page
  // arrives back on a page still stamped with the old one. The cookies are
  // the truth, so the root attributes and the controls' state are rewritten
  // from them, honouring only the values the server honours. An ordinary
  // load needs none of this and is gated out: the server just stamped the
  // same cookies itself.
  window.addEventListener('pageshow', (event) => {
    if (!event.persisted) {
      return;
    }
    const theme = readCookie('yomihon_theme');
    writeTheme(theme === 'dark' || theme === 'light' ? theme : null, false);
    const stored = readCookie('yomihon_textsize');
    const size = stored === 'l' || stored === 'xl' ? stored : 'm';
    root.dataset.textsize = size;
    textsizeToggle?.setAttribute('aria-label', textsizeLabel(size));
    // Only the literal "off" is off, matching the server's stamp: any other
    // cookie, or none, is the default on. Both the root and the control have
    // to move together or a restored page shows furigana the cookie refused
    // and a button that claims the opposite.
    const ruby = readCookie('yomihon_ruby') === 'off' ? 'off' : 'on';
    root.dataset.ruby = ruby;
    rubyToggle?.setAttribute('aria-pressed', String(ruby === 'on'));
    // The reading typeface honours the same three names the server does, and
    // like the theme it may have nothing to write: a reader who chose none
    // leaves the attribute off the root rather than on a default.
    const font = readCookie('yomihon_font');
    if (font === 'serif' || font === 'sans' || font === 'kai') {
      root.dataset.font = font;
    } else {
      delete root.dataset.font;
    }
    const shortcuts = readCookie('yomihon_shortcuts') === 'off' ? 'off' : 'on';
    root.dataset.singleKeyShortcuts = shortcuts;
    if (shortcutsToggle) shortcutsToggle.checked = shortcuts === 'on';
    // On the page where these choices are set, the radios say which one is in
    // force, so they are as much a stale claim as the attributes above.
    syncSettingsChoices();
    // The document's language cannot be rewritten in place, so a stale one
    // means asking for the page again. The comparison normalises the cookie
    // exactly as the server does — anything but "en" reads as the default —
    // so a value the server would ignore can never reload in a loop: the
    // fresh page always satisfies the same comparison.
    const lang = readCookie('yomihon_lang') === 'en' ? 'en' : 'zh-Hant';
    if (root.lang !== lang) {
      location.reload();
    }
  });

  // What the reader is looking at, and a way to be told when that changes.
  return {
    theme: effectiveTheme,
    onThemeChange(listener) {
      themeChanged = listener;
    },
  };
}

File: internal/wording/preferences.go
package wording

// The page a reader sets the reading on: one heading per choice, one sentence
// under each saying what it does and — where a reader would reasonably guess
// wider — what it leaves alone. The sentences are the point of the page. A
// list of switches with no words is a list a reader has to try one by one.

// PreferencesTitle names the page. PreferencesLink is the same word on the
// control that leads to it, kept separate because a heading and a link are
// free to part company later without either being renamed by accident.
var (
	PreferencesTitle = both("設定", "Settings")
	PreferencesLink  = both("設定", "Settings")
)

// PreferencesLede is the whole promise of the page in one line: nothing chosen
// here is written into a note, and nothing here travels.
var PreferencesLede = both(
	"這些選擇只存在這台瀏覽器裡，不進筆記、也不出這台機器。",
	"These choices live in this browser alone — never in a note, never off this machine.",
)

// The interface language, and the boundary a reader would otherwise have to
// discover: the words yomihon writes change, the words an author wrote do not.
var (
	PrefLanguage     = both("介面語言", "Language")
	PrefLanguageNote = both(
		"只換介面的字；筆記保留作者寫的語言。",
		"Switches the interface only; a note keeps the language it was written in.",
	)
)

// The two languages name themselves in themselves. A reader looking for the
// way into a language they can read needs to recognise its name, not read a
// translation of it in the language they are leaving.
var (
	PrefLanguageZh = both("繁體中文", "繁體中文")
	PrefLanguageEn = both("English", "English")
)

// The light and dark grounds, and what choosing neither means.
var (
	PrefAppearance     = both("外觀", "Appearance")
	PrefAppearanceNote = both(
		"跟隨系統時，深淺由作業系統決定。",
		"Following the system leaves light and dark to the operating system.",
	)
	PrefAppearanceSystem = both("跟隨系統", "Follow the system")
	PrefAppearanceLight  = both("亮色", "Light")
	PrefAppearanceDark   = both("暗色", "Dark")
)

// The reading column's measure. The note draws the line the control does not:
// the article grows, the header and rails around it hold still.
var (
	PrefTextSize           = both("字級", "Text size")
	PrefTextSizeMedium     = both("中", "Medium")
	PrefTextSizeLarge      = both("大", "Large")
	PrefTextSizeExtraLarge = both("特大", "Extra large")
	PrefTextSizeNote       = both(
		"只放大內文，介面不變。",
		"Enlarges the reading column only; the header and rails around it keep their size.",
	)
)

// The face the reading is set in, and the three it may be set in. Each side
// names them as its own readers know them: the Chinese page by the characters,
// the English page by the names those faces carry in English.
var (
	PrefTypeface     = both("字體", "Typeface")
	PrefTypefaceNote = both(
		"只換內文與課文練習的字體；標題、介面、程式碼不變。",
		"Changes the reading column and the lesson exercises only; headings, the interface, and code keep theirs.",
	)
	PrefTypefaceSerif = both("明體", "Serif")
	PrefTypefaceSans  = both("黑體", "Sans serif")
	PrefTypefaceKai   = both("楷體", "Kaiti")
)

// The reading aids over Japanese text. The note says the switch is inert
// everywhere else, so a reader who turns it on and sees nothing knows why.
var (
	PrefFurigana     = both("顯示讀音", "Readings")
	PrefFuriganaNote = both(
		"只影響日文上方標了讀音的頁面。",
		"Applies to the pages that carry readings above the Japanese text.",
	)
)

// Whether a bare key does anything, and which keys the answer covers. Naming
// the two that keep working is the whole value of the sentence: "off" reads as
// "no keyboard" without it.
var (
	PrefShortcuts     = both("單鍵快捷鍵", "Single-key shortcuts")
	PrefShortcutsNote = both(
		"關掉的是 / 與 [；⌘K 與 Esc 不受影響。",
		"Turns off / and [; ⌘K and Esc keep working.",
	)
)

// The two states every switch on this page is in. They are the words a reader
// picks, so they are verbs' worth of plain rather than the glyphs the header
// controls carry.
var (
	PrefOn  = both("開啟", "On")
	PrefOff = both("關閉", "Off")
)

// The one control that is not a choice between values. Its sentence has to
// name the whole of what it clears, including the consequence a reader would
// not predict: the interface comes back speaking the language it started in.
var (
	PrefReset     = both("回到預設", "Back to the defaults")
	PrefResetNote = both(
		"清掉這一頁的每一項設定（含介面語言，所以套用後介面回到繁體中文）；本次分頁的側欄狀態與篩選字不受影響。",
		"Clears every setting on this page — including the interface language, so the page comes back in Traditional Chinese; this tab's sidebar state and filter text are untouched.",
	)
)

// PrefApply is on the control that saves the page where nothing is running to
// save it as the reader picks. It is drawn on every page and hidden where the
// enhancement is live, so a browser without scripting sees the control it
// needs and a browser with it never sees a control it does not.
var PrefApply = both("套用", "Apply")

// PrefReturn leads back to whatever the reader was holding when they came
// here. It stands in for the apply control where choices are already stored as
// they are picked: without it the page would be the one place that saves
// everything and offers no way out of itself.
var PrefReturn = both("回到剛才在讀的", "Back to what you were reading")

// PrefSaveRefused is what a choice says for itself when the browser would not
// keep it. The choice is put back to what it was, because a radio left sitting
// on a value nothing stored is the page telling the reader something untrue —
// and the likeliest cause is named, since it is one a reader can act on.
var PrefSaveRefused = both(
	"這個選擇沒有存下來，已經放回原本的值。這個瀏覽器可能不讓這個位址存 cookie。",
	"That choice was not stored, and has been put back. This browser may be refusing cookies for this address.",
)

// What this browser is holding, said plainly and completely. A page that
// stores things on a reader's machine should be the page that lists them.
var (
	PrefStorageTitle = both("這個瀏覽器記著什麼", "What this browser remembers")
	PrefStorageNote  = both(
		"設定存成這個位址上的 cookie，側欄的狀態只活到分頁關閉；沒有任何一項離開這台機器。",
		"Settings are cookies on this address; the sidebar's own values die with the tab; none of it leaves this machine.",
	)
	PrefStorageCookies = both(
		"cookie 記著：介面語言、外觀、字級、字體、顯示讀音、單鍵快捷鍵",
		"Cookies: language, appearance, text size, typeface, readings, single-key shortcuts",
	)
	PrefStorageSession = both(
		"分頁內的值：側欄展開狀態、側欄篩選字",
		"Per-tab values: sidebar expansion, sidebar filter text",
	)
)

// What the page says to a request the rendered forms could not have sent. The
// forms offer one field carrying one of the values named above, so neither
// sentence appears in ordinary use.
var (
	PrefUnknownValue = both(
		"這不是這個設定認得的值，沒有任何東西被改動。",
		"That is not a value this setting knows; nothing was changed.",
	)
	PrefFormUnreadable = both("表單讀不出來。", "The form could not be read.")
)
